enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,316 @@
+
+##### API load balancer migration
+
+Note: API LB migration is a significant infrastructure change that may cause an API outage for some time. There is no requirement for the process to take place without service disruption.
+
+-- self-hosted to an external Load balancer
+
+In order to migrate the self-hosted LB to an external Load balancer the admin should:
+ - Provide new IP address (!= API-VIP from install-config) pointing to external Load balancer front end.
+ - Edit apiintIpAddress value to use this IP address.
+ 
+As a result of this change the CHNSO should:
+- First, verify that IP address is valid and kube-api is available using the new IP address.
+- Update CoreDNS to resolve api-int to the new IP address
+- Remove the HAProxy pods after graceful shutdown timeout 
+- Keepalived static pods will continue running
+
+-- external to self-hosted Load balancer 
+
+To re-activate the self-hosted Load balancer the admin should set apiintIpAddress value to API-VIP address provided in install-config file and verify that external Load balancer is operational for graceful timeout.
+
+Post this change the CHNSO should run the same procedure described above and run the HAProxy pods.
+
+
+### User Stories
+
+#### As an admin I want to disable the self-hosted on-prem LB and DNS stack.
+
+#### As an admin I want to use the self-hosted on-prem LB and DNS stack also in UPI deployments.
+
+#### As an admin I want to retrieve the self-hosted stack status through K8S API.
+
+
+### Test Plan
+The operator will be tested in the OpenShift CI via e2e tests triggered using the [e2e-metal-ipi](https://github.com/openshift/release/blob/master/ci-operator/step-registry/baremetalds/e2e/baremetalds-e2e-workflow.yaml) workflow defined in the OpenShift CI steps registry and also by e2e tests for openstack, ovirt and vsphere.
+
+Additionally, the following tests should be added:
+- Check each platform when external Load balancer is used
+- Disruptive test that switch on/off the self-hosted components
+- Check Kube-api downtime (the values ​​should be similar to the values ​​when using static pods) for master node reboot and node drain cases.
+
+### Upgrade / Downgrade Strategy
+
+Prior to this proposal, the self-hosted stack implementation was running through the MCO.
+The implementation of the self-hostd stack using the cluster-hosted-net-services operator might lead to cases during upgrade/downgrade in which two parallel components are running the self-hosted stack. 
+ 
+#### Design
+
+The cluster-hosted-net-services operator will:
+1. Use the same code used by MCO to generate configuration files of HAProxy CoreDNS in each node.
+2. Run the CoreDNS and HAProxy instances in each node following verification that corresponding instance created by MCO isn't running anymore in this node.
+
+The implementation of 1 and 2 will ensure that LB and DNS services will be available in case that both MCO and cluster-hosted-net-services operator run the self-hosted stack.
```

Comment:
This section could talk more about the hand-off during an upgrade. What is responsible for deleting the old static pods that are being converted to the DaemonSet for things like CoreDNS-MDNS and MDNS-publisher? How do we handle version skew between the MCO trying to manage those things and the new operator?

Reply:
If a release includes the cluster-hosted-network-service operator that means it includes MCO that shouldn't run HAProxy CoreDNS-MDNS and MDNS-publisher static pods. 
So, after upgrade to this release MCO is responsible to delete these static pods.
I'll update this section

Reply:
Sure, but upgrades aren't atomic across the cluster, so there's a very good chance that the old MCO and new operator will be running at the same time. We had to deal with this in the hand-off between the machine-api-operator and cluster-baremetal-operator by making the N-1 version of MAO understand that it might encounter the CBO at some point. There are more details about the approach we took in that enhancement doc at https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#upgrade--downgrade--version-skew-strategy

Reply:
> So, after upgrade to this release MCO is responsible to delete these static pods.

I wouldn't assume this will happen. I don't believe MCO removes files that are no longer present in its templates. We may need to add something to the operator that removes the old static pod definitions.

I don't think we'll have a lot of trouble with the handoff from MCO static pods to operator daemonsets though. When the daemonset comes up it will fail because the static pod will be holding the port. When the static pods are removed then the port will open up and the daemonset will be able to start properly. Since the functionality of the services should be equivalent between MCO and the new operator it won't matter which is running, as long as at least one is. We do need to test that and verify there isn't a gap between when the static pods are removed and the operator ones start.

Reply:
So the new operator is completely responsible for the migration from static pods to DaemonSet? Should it wait for the MCO to be upgraded so the MCO won't re-apply the static pod definition?

What happens if all of the nodes happen to do that migration at the same time and the services being managed are offline for a while? Do we need to roll the migration out across hosts one at a time?

Reply:
> So the new operator is completely responsible for the migration from static pods to DaemonSet? Should it wait for the MCO to be upgraded so the MCO won't re-apply the static pod definition?

I believe so, although Yossi is going to confirm that MCO works the way I think it does. If that's the case, I don't think we have to worry about when MCO is upgraded because the new version won't have those static pods so it won't re-render them. That said, currently the operator is ordered such that it will update after MCO anyway, which is probably safest.

> What happens if all of the nodes happen to do that migration at the same time and the services being managed are offline for a while? Do we need to roll the migration out across hosts one at a time?

We do need to be a little careful with keepalived at least so we don't lose the API VIP. I think the simplest thing might be to add an init container that will delete any old static pods right before the new service starts up. Even if all the pods happened to deploy at the same time and all of the old keepaliveds get deleted, the new ones will be ready to take over immediately.

Reply:
> Sure, but upgrades aren't atomic across the cluster, so there's a very good chance that the old MCO and new operator will be running at the same time. We had to deal with this in the hand-off between the machine-api-operator and cluster-baremetal-operator by making the N-1 version of MAO understand that it might encounter the CBO at some point. There are more details about the approach we took in that enhancement doc at https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#upgrade--downgrade--version-skew-strategy

AFAIU, the use case here is a bit different than the CBO and MAO, 
in CBO, MAO case:
- The CRD resource already exist before release 4.N 
- It was managed by MAO in release <= 4.N-1
- It should be managed by CBO in release >= 4.N

So, there's a need to sync between CBO and MAO such that only a single component should manage the resource.
While in our case, the CRD will be introduced only in release 4.N, and the MCO isn't aware of this resource. 

Though there is no need to sync between MCO and new operator which component should handle the CRD, we should make sure there's always a running instance of HAProxy, CoreDNS and mDNS-publisher in each node 

We plan to do that by :
A. Upgrade the operator before the MCO (by setting operator's manifests lower priority than MCO priority )
B. Operator's Daemonset pods verification that the corresponding instance created by MCO isn't running anymore in this node.  That should be easy to implement since both static pods and daemonset pods run on host networking and listen on a well-known port.

Reply:
> I wouldn't assume this will happen. I don't believe MCO removes files that are no longer present in its templates. We may need to add something to the operator that removes the old static pod definitions.

I checked that, MCO deletes the unneeded static pod (HAProxy, coredns, mdns-publisher) manifests, so there's no need for the operator to take care of that.
MCO doesn't delete the conf files (e.g: haproxy.conf) used by the static pods, but since operator's daemonset pods don't use host mounting for conf files - it should not affect them. I think we should add code to MCO (release 4.N) to verify that the unused (e.g: /etc/coredns/ ) host mounted directories being removed.

Reply:
> We do need to be a little careful with keepalived at least so we don't lose the API VIP. I think the simplest thing might be to add an init container that will delete any old static pods right before the new service starts up. Even if all the pods happened to deploy at the same time and all of the old keepaliveds get deleted, the new ones will be ready to take over immediately.

Keepalived remains as static pod under MCO, I don't think we can lose API-VIP and we don't need special treatment for this case.
The upgrade of MCO requires node reboot and it performs node after node, so there's no case in which all nodes being rebooted at the same time.

Reply:
> So the new operator is completely responsible for the migration from static pods to DaemonSet? Should it wait for the MCO to be upgraded so the MCO won't re-apply the static pod definition?

No, we should have the following steps  during upgrade in each node:
0. MCO (release N-1) still run components as static pods 
1. Operator (release N) starts to run and deploy Daemonset pods (HAProxy, coreDNS, etc)
    1.1 Each one of these DS pods should wait till the corresponding static pod is deleted before it starts running 
2. post Node reboot the MCO upgraded to release N (all the unneeded static pods removed)
3. DS pods should take control

Reply:
> AFAIU, the use case here is a bit different than the CBO and MAO,
> in CBO, MAO case:
> 
> * The CRD resource already exist before release 4.N
> * It was managed by MAO in release <= 4.N-1
> * It should be managed by CBO in release >= 4.N
> 
> So, there's a need to sync between CBO and MAO such that only a single component should manage the resource.
> While in our case, the CRD will be introduced only in release 4.N, and the MCO isn't aware of this resource.

The coordination was necessary to ensure that only 1 controller was responding to the API so the different implementations did not fight with each other. In your case the CRD may be new, but the services being managed behind that CRD are still there, and there can still be conflicts.

> Though there is no need to sync between MCO and new operator which component should handle the CRD, we should make sure there's always a running instance of HAProxy, CoreDNS and mDNS-publisher in each node
> 
> We plan to do that by :
> A. Upgrade the operator before the MCO (by setting operator's manifests lower priority than MCO priority )

Manifest priority lets you control when a component starts upgrading because it tells you what order the CVO is going to make API calls. It does not ensure that one component is fully upgraded before the next is started, though, because after the CVO does its work the operators have to launch and do theirs, and that happens in parallel. So, we effectively have no way to guarantee the order of operations during an upgrade.

> B. Operator's Daemonset pods verification that the corresponding instance created by MCO isn't running anymore in this node. That should be easy to implement since both static pods and daemonset pods run on host networking and listen on a well-known port.

That helps if the new operator starts the DaemonSet on the pod before the MCO starts making its changes. If the MCO rolls out its changes first, there will be an indeterminate amount of time between when the static pod is taken down and the DaemonSet is launched. How big of a gap can we afford to have? What impact will there be on the cluster if one or more nodes do not have the required services running?

Reply:
That's a good point.
I think it should be fine even if MCO rolls out first, in that case, the node should be able to join the cluster because Keepalived (will provide the api-vip) and dnsmasq (will resolve api-int name to api-vip) still run by MCO.